sympy/stats/tests/test_stochastic_process.py

```diff
@@ -86,24 +95,79 @@ def test_DiscreteMarkovChain():
     Y2 = DiscreteMarkovChain('Y', trans_probs=TO2)
     Y3 = DiscreteMarkovChain('Y', trans_probs=TO3)
     assert Y3._transient2absorbing() == None
-    raises (ValueError, lambda: Y3.fundamental_matrix())
     assert Y2.is_absorbing_chain() == True
     assert Y3.is_absorbing_chain() == False
+
+    assert not Y2.is_irreducible
+    assert Y3.is_irreducible
+
+    p = Y2.stationary_distribution()
+    l = Y2.limiting_distribution()
+    assert p * TO2 == p == l
+    p = Y3.stationary_distribution()
+    l = Y3.limiting_distribution()
+    assert p * TO3 == p == l
+    assert Y2.communication_classes() == [CommunicationClass([0], True, 1),
+                                          CommunicationClass([1, 2], False, 1)]
+    assert Y3.communication_classes() == [CommunicationClass([0, 1, 2], True, 1)]
+    assert Y2.canonical_form() == ([0, 1, 2], TO2)
+    assert Y3.canonical_form() == ([0, 1, 2], TO3)
+    assert Y2.decompose() == ([0, 1, 2], TO2[0:1, 0:1], TO2[1:3, 0:1], TO2[1:3, 1:3])
+    assert Y3.decompose() == ([0, 1, 2], TO3, Matrix(0, 3, []), Matrix(0, 0, []))
+
+    assert Y2.fundamental_matrix() == Matrix([[3, 4], [3, 8]])
+    assert Y3.fundamental_matrix() == Matrix(0, 0, [])
+    assert Y2.limiting_transient_matrix() == Y2.fundamental_matrix()*TO2[1:3, 0:1] == Y2.absorbing_probabilites()
+    assert Y3.limiting_transient_matrix() == Y3.fundamental_matrix()*Matrix(0, 3, []) == Y3.absorbing_probabilites()
+
+    assert Y2.expected_time_to_absorption() == Matrix([[6], [10]])
+    assert Y3.expected_time_to_absorption() == Matrix(0, 1, [])
+
+    assert Y2.first_passage_matrix(1) == Y2.transition_probabilities
+    assert Y2.first_passage_matrix(2) == Matrix([[0, 0, 0], [S(1)/9, S(1)/12, S(1)/9], [S(1)/12, S(3)/16, S(1)/12]])
+    assert Y3.first_passage_matrix(1) == Y3.transition_probabilities
+    assert Y3.first_passage_matrix(2) == Matrix([[S(1)/4, S(3)/16, S(1)/4], [S(1)/9, S(1)/3, S(1)/9], [S(1)/12, S(3)/16, S(1)/12]])
+
+    assert Y2.first_passage_matrix(2, 2, 1) == S(3)/16
+    assert Y2.first_passage_matrix(2, 1, 1) == S(1)/12
+    assert Y3.first_passage_matrix(2, 2, 1) == S(3)/16
+    assert Y3.first_passage_matrix(2, 1, 1) == S(1)/3
+    assert (Y2.first_passage_matrix(t, 1, 2) - 3**-t).simplify() == 0
+
     TO4 = Matrix([[Rational(1, 5), Rational(2, 5), Rational(2, 5)], [Rational(1, 10), S.Half, Rational(2, 5)], [Rational(3, 5), Rational(3, 10), Rational(1, 10)]])
     Y4 = DiscreteMarkovChain('Y', trans_probs=TO4)
     w = ImmutableMatrix([[Rational(11, 39), Rational(16, 39), Rational(4, 13)]])
-    assert Y4.limiting_distribution == w
+    assert Y4.limiting_distribution() == w
```

Comment:
I'd suggest not to change public API unless really necessary. Are there any significant gains or is this just a result of design decision?

Reply:
Sometimes the limiting distribution relies on the initial distribution, `p0`. This needs to be an argument in that case. For example, with the transition matrix = `Matrix([[1, 0], [0, 1]])`, the limiting distribution is `p0`.

Reply:
Maybe I could rename the limiting distribution to something else. Then `limiting_distribution` will just call that new function with default arguments.

Reply:
Fixed. `limiting_distribution` now calls `limiting_dist()` with default arguments.